## 📘 Introducción

En este notebook se abordará el desarrollo de modelos de regresión para predecir el consumo de energía de edificaciones, en función de características relacionadas con su diseño y configuración.

Utilizaremos diversos algoritmos de machine learning para comparar su desempeño y posteriormente optimizarlos. Se evaluará cada modelo en términos de métricas de error y capacidad predictiva, buscando un equilibrio entre precisión y generalización.

Este flujo de trabajo seguirá los siguientes pasos:

*  Carga y preparación de los datos  
*  Separación en conjunto de entrenamiento y prueba  
*  Entrenamiento de modelos base (baseline)  
*  Optimización de hiperparámetros en distintos modelos
*  Evaluación comparativa  
*  Análisis de resultados y selección del modelo final


# Librerías y Carga de los datos

In [18]:
# Librerías necesarias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

from sklearn.linear_model import Ridge

In [13]:
X = pd.read_csv('../data/features.csv')
y1 = pd.read_csv('../data/y_heating.csv')
y2 = pd.read_csv('../data/y_cooling.csv')

# Renombrar columnas para mayor claridad (según documentación UCI)
X.columns = [
    'Relative_Compactness', 'Surface_Area', 'Wall_Area', 'Roof_Area',
    'Overall_Height', 'Orientation', 'Glazing_Area',
    'Glazing_Area_Distribution']

y1.columns = ['Heating_Load']
y2.columns = ['Cooling_Load']

In [14]:
# División de datos: usaremos 80% entrenamiento y 20% test
X_train, X_test, y_train, y_test = train_test_split(
    X, y1, test_size=0.2, random_state=42
)

print("Train shape:", X_train.shape)
print("Test shape:", X_test.shape)

Train shape: (614, 8)
Test shape: (154, 8)


# Baseline

## 🔍 4. Modelos Baseline

Para establecer una línea base de rendimiento, en esta sección implementaremos un **modelo de regresión Ridge**, el cual será nuestro punto de comparación inicial.

Este modelo es una versión regularizada de la regresión lineal, que introduce una penalización sobre los coeficientes del modelo, lo que ayuda a prevenir el sobreajuste en presencia de multicolinealidad o muchas variables independientes.

En la siguiente sección se explorarán y optimizarán otros modelos más complejos, como:
- Árboles de Decisión
- Random Forest
- K-Nearest Neighbors (KNN)

### 🔧 Métricas utilizadas

Evaluaremos el modelo con las siguientes métricas:

- **MAE** (Mean Absolute Error): Promedio del error absoluto, fácil de interpretar.
- **RMSE** (Root Mean Squared Error): Penaliza más fuertemente los errores grandes.
- **R² Score**: Cuantifica la proporción de la varianza explicada por el modelo.


In [21]:
# Escalado de características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Entrenamiento del modelo Ridge
ridge = Ridge()
ridge.fit(X_train_scaled, y_train)

# Predicciones
y_pred = ridge.predict(X_test_scaled)

In [22]:
# Evaluación
mae = mean_absolute_error(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred, squared=False)
r2 = r2_score(y_test, y_pred)

# Resultados
print("Evaluación del modelo Ridge (baseline):")
print(f"MAE:  {mae:.4f}")
print(f"RMSE: {rmse:.4f}")
print(f"R²:   {r2:.4f}")

Evaluación del modelo Ridge (baseline):
MAE:  2.1908
RMSE: 3.0354
R²:   0.9116


### 📌 Observaciones

El modelo de regresión Ridge ha logrado un buen desempeño inicial considerando que no se ha realizado ajuste de hiperparámetros. 

- El **MAE de 2.19** indica que, en promedio, las predicciones se desvían en torno a 2.2 unidades del valor real.
- El **RMSE de 3.03**, al ser ligeramente mayor al MAE, sugiere que existen algunos errores más grandes, pero sin ser excesivamente severos.
- El **R² de 0.91** muestra que el modelo explica más del 90% de la varianza de la variable objetivo, lo cual es un muy buen punto de partida.

Este resultado servirá como referencia para evaluar si modelos más complejos o ajustados pueden superar esta línea base.
